[Pathways] Create CountOverTimeWithAvg metric, chart, and viz

Catacola opened the pull request:
## Description of the change

Creates a CountOverTimeWithAvg chart and viz.

Creates `SupervisionTransitionCountOverTimeMetric` and pipes it into the above viz. Populates the metric with supervision to prison data. Data for development is in the staging bucket's ID folder. You can also use fixture files for offline development.

The chart is built using the component for the practices summary charts.

NOTE: The chart still needs a lot of styling work, but it displays the data correctly. For example, it is showing counts and percentages and it's not formatting the date axis correctly. However, this will allow for development of additional metrics and pages.

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Configuration change (adjusts configuration to achieve some end related to functionality, development, performance, or security)

## Related issues

Part of #1318 

## Checklists

### Development

These boxes should be checked by the submitter prior to merging:

- [x] Manual testing against realistic data has been performed locally
- [ ] E2E have been run for Lantern changes ([Steps in the Readme](https://github.com/Recidiviz/pulse-dashboard#running-e2e-tests))

### Code review

These boxes should be checked by reviewers prior to merging:

- [x] This pull request has a descriptive title and information useful to a reviewer
- [x] This pull request has been moved out of a Draft state, has no "Work In Progress" label, and has assigned reviewers
- [x] Potential security implications or infrastructural changes have been considered, if relevant

nasaownsky commented:
Hey @jovergaag @Catacola! Thanks a lot for helping me out with this, great work you both! There's still lot of work to do, but we moving in right direction. 

Few notes:

1. I see that @Catacola reused all `PracticesSummary` folder, but I propose to leave it untouched, since this chart and `ChartCountOverTimeWithAvg` are alike but slightly different.

2. What `sectionId` should have `supervisionToPrison` and `supervisionToLiberty`? If Prison and Supervision has `populationOverTime` should it be `transitionCountOverTime` or something?

jovergaag commented on src/core/models/SupervisionTransitionCountOverTimeMetric.ts:
```diff
@@ -0,0 +1,120 @@
+  }
+
+  get note(): string {
+    return `${this.noteCopy}`;
```
This is already on `PathwaysMetric` so we don't need to override it here.

jovergaag commented on src/core/ChartCountOverTimeWithAvg/ChartCountOverTimeWithAvg.tsx:
```diff
@@ -0,0 +1,190 @@
+}) => {
+  const [hoveredId, setHoveredId] = useState(null);
+
+  if (isLoading || !data) return <div className="ChartCountOverTimeWithAvg" />;
```
The `ModelHydrator` takes care of the loading state, so this probably only needs to check for data. The styling on this isn't quite right when there isn't data, but we have a ticket queued up for no data state so we can just leave this until then!

jovergaag commented on src/core/ChartCountOverTimeWithAvg/ChartCountOverTimeWithAvg.tsx:
```diff
@@ -0,0 +1,190 @@
+        rAccessor="value"
+        rExtent={yRange}
+        oLabel={(date: string, _: any, index: number) => {
+          // Display the first and then every 30 labels
```
We seem to be getting only one x-axis label. This first and then ever 30 labels probably doesn't apply to the shorter time periods.

Catacola replied:
Thanks for pointing that out. If we keep this as one component, I'll add a prop controlling how often the x-axis is labeled.

jovergaag commented on src/core/PathwaysTooltip/PathwaysTooltip.tsx:
```diff
@@ -0,0 +1,54 @@
+  };
+  return (
+    <div className="PathwaysTooltip" style={transformStyle}>
+      <div className="PathwaysTooltip__Date">{date}</div>
```
Same here on the BEM lower case naming convention for objects/modifiers

Catacola replied:
Copied form an old component. Will fix.

soniachokshi commented:
Looks great, team! Did a quick design review, and wrote up my comments [here](https://www.figma.com/file/Mw6kSEP2jdr9i7UVEjPuCI/UP-Consolidation?node-id=2248%3A389) in Figma. Let me know if you have any questions.

jovergaag commented on src/core/PopulationTimeSeriesChart/PopulationTimeSeriesLegend.tsx:
```diff
@@ -28,7 +28,12 @@ const PopulationTimeSeriesLegend: React.FC<propTypes> = ({ items }) => {
         <div className="PopulationTimeSeriesLegend__Item" key={label}>
-          <div className={`PopulationTimeSeriesLegend__Icon ${label}Legend`}>
+          <div
+            className={`PopulationTimeSeriesLegend__Icon ${label.split(
```
Same with the BEM naming in this component as well, though clearly some capitals have already made it in!

jovergaag commented on src/core/PracticesSummaryChart/PracticesSummaryTooltip.tsx:
```diff
@@ -17,8 +17,9 @@
 import React from "react";
 
-import { formatISODateString, formatPercent } from "../../utils/formatStrings";
+import { formatISODateString, formatPercent } from "../../utils";
```
I don't see this `PracticesSummaryTooltip` being used anywhere in the app, which means that `PathwaysTooltip` is also not being used.

jovergaag commented on src/core/PathwaysTooltip/PathwaysTooltip.tsx:
```diff
@@ -0,0 +1,54 @@
+/*
```
I don't see where this component is actually being used. It is used in a different tooltip component, that, itself is not used.

Catacola replied:
The later component used to be used in the bar chart. This refactor was to decouple the styling of the tooltip from its contents so that we could have custom tooltip text per chart. I just didn;t manage to implement it last night.

jovergaag commented on src/core/ChartCountOverTimeWithAvg/ChartCountOverTimeWithAvg.tsx:
```diff
@@ -0,0 +1,190 @@
+        }}
+        tooltipContent={(d: any) => {
+          const pieceData = d.pieces[0];
+          return <PopulationTimeSeriesTooltip d={pieceData} />;
```
I'm getting an error from the tooltips on the Practices page when rolling over the chart.

```
index.js:373 Uncaught RangeError: Invalid time value
    at format (index.js:373)
    at formatDate (formatStrings.ts:195)
    at PopulationTimeSeriesTooltip (PopulationTimeSeriesTooltip.tsx:43)
    at renderWithHooks (react-dom.development.js:14985)
    at mountIndeterminateComponent (react-dom.development.js:17811)
    at beginWork (react-dom.development.js:19049)
    at HTMLUnknownElement.callCallback (react-dom.development.js:3945)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994)
    at invokeGuardedCallback (react-dom.development.js:4056)
    at beginWork$1 (react-dom.development.js:23964)
    at performUnitOfWork (react-dom.development.js:22779)
    at workLoopSync (react-dom.development.js:22707)
    at renderRootSync (react-dom.development.js:22670)
    at performSyncWorkOnRoot (react-dom.development.js:22293)
    at react-dom.development.js:11327
    at unstable_runWithPriority (scheduler.development.js:646)
    at runWithPriority$1 (react-dom.development.js:11276)
    at flushSyncCallbackQueueImpl (react-dom.development.js:11322)
    at flushSyncCallbackQueue (react-dom.development.js:11309)
    at batchedEventUpdates$1 (react-dom.development.js:22403)
    at batchedEventUpdates (react-dom.development.js:3745)
    at dispatchEventForPluginEventSystem (react-dom.development.js:8507)
    at attemptToDispatchEvent (react-dom.development.js:6005)
    at dispatchEvent (react-dom.development.js:5924)
    at unstable_runWithPriority (scheduler.development.js:646)
    at dispatchUserBlockingUpdate (react-dom.development.js:5894)
```

jovergaag replied:
Did you intend to change this to `PathwaysTooltip`?

nasaownsky replied:
This is because I changed chart Date logic, and since this chart is reused in Practice view in PracticesSummaryChart - the error occurs there. We need either divide this charts (as I has propose) or figure out something else.

Catacola replied:
Before I pushed this PR off last night, I was playing with the ability to pass in the tooltip as a prop to the chart. This would allow us to customize the tooltip to its specific use. However, I ran out of energy before I was able to implement it. (This also explains the phantom tooltip component.)

As for splitting the charts, I'm fine with that since the new one will be used in plenty of other places in the new pathways view. I may have been stretching for commonalities where there were none.

jovergaag commented:
> 1. I see that @Catacola reused all `PracticesSummary` folder, but I propose to leave it untouched, since this chart and `ChartCountOverTimeWithAvg` are alike but slightly different.

This is a good point. Most of the difference between `PracticesSummaryChart` and its child `ChartCountOverTimeWithAvg` is just some data transformations. We could move that transformations to the metric and just render `ChartCountOverTimeWithAvg` directly on `PagePractices`. Up to you @Catacola

jovergaag commented:
> 2\. What `sectionId` should have `supervisionToPrison` and `supervisionToLiberty`? If Prison and Supervision has `populationOverTime` should it be `transitionCountOverTime` or something?

This is another great point! It would be nice to maintain a consistent sectionId for all of the "over time" charts. But `population` does not accurately describe the transition counts. Maybe `countOverTime` is a better id. Is that too impersonal considering it will be public facing in the path @Catacola

jovergaag commented on src/core/models/SupervisionTransitionCountOverTimeMetric.ts:
```diff
@@ -0,0 +1,120 @@
+import { SupervisionTransitionTimeSeriesRecord } from "./types";
+import { getRecordDate } from "./utils";
+
+export default class SupervisionTransitionCountOverTimeMetric extends PathwaysMetric<SupervisionTransitionTimeSeriesRecord> {
```
Now that I've been through all of this, I'm wondering if a more generic name would be better here? `CountOverTimeWithAvgMetric`? Or do we like the descriptive naming to make it easier to quickly grok what the metric is for?

jovergaag replied:
After having thought about it a bit more, maybe we do need to make it more generic since our next few states will not have population projections, so supervision and prison will render this count over time chart instead of the projections chart.

Catacola replied:
Sure thing. I'll play around with the name. One thing about it is that it is supervision-specific since it has `supervision_level`. We may want two generic over time metrics: one for supervision populations and one for prison populations. This may end up being a job for traits.

jovergaag commented on src/core/models/utils.ts:
```diff
@@ -39,3 +41,28 @@ export function createProjectionTimeSeries(
+  rawRecords: RawMetricData,
+  countField: string
+): SupervisionTransitionTimeSeriesRecord[] {
+  return rawRecords.map((record) => {
```
We may want to add a sort onto this, with a note to remove the sort once the data is sorted in the view.

jovergaag commented on src/core/ChartCountOverTimeWithAvg/ChartCountOverTimeWithAvg.tsx:
```diff
@@ -0,0 +1,190 @@
+  );
+};
+
+export default observer(ChartCountOverTimeWithAvg);
```
We don't need `observer` here.

Catacola replied:
Hah, you're right! I did remove all the stores.

Catacola commented on src/core/PathwaysTooltip/PathwaysTooltip.scss:
```diff
@@ -0,0 +1,36 @@
+.PathwaysTooltip {
+  @include chart-tooltip;
+
+  &__Date,
```
I can clean this up. This was the existing style. It must have been missed in the first review.

Catacola commented on src/core/VizCountOverTimeWithAvg/VizCountOverTimeWithAvg.tsx:
```diff
@@ -0,0 +1,54 @@
+  metric: SupervisionTransitionCountOverTimeMetric;
+};
+
+const VizCountOverTimeWithAvg: React.FC<VizCountOverTimeWithAvgProps> = ({
```
`ChartCountOverTimeWithAvg` was created by pulling the Practices summary chart into its own component. `ChartCountOverTimeWithAvg` is used in the `PracticesSummaryChart` component directly since that component has access to the PracticesStore and just pulls the data directly. Wrapping `ChartCountOverTimeWithAvg` in `VizCountOverTimeWithAvg` just allows for the unpacking of the metric here.

Catacola commented:
I've split the practices plot back out into its original component and cleaned up the one we're using for Pathways. I also created a new section, `transitionsOverTime`.

jovergaag commented on src/core/content/page/default.ts:
```diff
@@ -20,6 +20,7 @@
 
 import { PageCopy } from "../types";
 
+// TODO(#1340): [Pathways] Get sections definition into spreadsheet or remove from content system
```
Good call!

jovergaag reviewed (approved):
woot!

jovergaag commented:
Huge team effort on this PR. Nice job team @Catacola @nasaownsky 💪🏻
